ambry-router/src/test/java/com.github.ambry.router/MockServer.java

```diff
@@ -53,61 +73,145 @@
-                dStream.writeShort(versionId);
-                dStream.writeInt(correlationId);
-                dStream.writeInt(0); // avoiding adding clientId
-                dStream.writeShort((short) error.ordinal());
-                return ByteBuffer.wrap(bStream.toByteArray());
-              }
-            }.getPayload());
+        response = makeDeleteResponse((DeleteRequest) send, serverError);
+        break;
       default:
         throw new IOException("Unknown request type received");
     }
+    ByteBufferChannel channel = new ByteBufferChannel(ByteBuffer.allocate((int) response.sizeInBytes()));
+    response.writeTo(channel);
+    ByteBuffer payload = channel.getBuffer();
+    payload.flip();
+    // read the size off. the size is used by the client to allocate the buffer and then interpret the response.
+    // MockServer abstracts it away at a level above that and returns the "allocated buffer".
+    payload.getLong();
+    return new MockBoundedByteBufferReceive(payload);
+  }
+
+  PutResponse makePutResponse(PutRequest putRequest, ServerErrorCode putError)
+      throws IOException {
+    if (putError == ServerErrorCode.No_Error) {
+      updateBlobMap(putRequest);
+    }
+    return new PutResponse(putRequest.getCorrelationId(), putRequest.getClientId(), putError);
+  }
+
+  GetResponse makeGetResponse(GetRequest getRequest, ServerErrorCode getError)
+      throws IOException {
+    GetResponse getResponse;
+    if (getError == ServerErrorCode.No_Error) {
+      List<PartitionRequestInfo> infos = getRequest.getPartitionInfoList();
+      if (infos.size() != 1 || infos.get(0).getBlobIds().size() != 1) {
+        getError = ServerErrorCode.Unknown_Error;
+      }
+    }
+
+    ServerErrorCode serverError;
+    ServerErrorCode partitionError;
+    // getError could be at the server level or the partition level. For partition level errors,
+    // handle set it in the partitionResponseInfo
+    if (getError == ServerErrorCode.No_Error || getError == ServerErrorCode.Blob_Expired
+        || getError == ServerErrorCode.Blob_Deleted || getError == ServerErrorCode.Blob_Not_Found) {
+      partitionError = getError;
+      serverError = ServerErrorCode.No_Error;
+    } else {
+      serverError = getError;
+      // does not matter.
+      partitionError = ServerErrorCode.No_Error;
+    }
+
+    if (serverError == ServerErrorCode.No_Error) {
+      int byteBufferSize;
+      ByteBuffer byteBuffer;
+      StoreKey key = getRequest.getPartitionInfoList().get(0).getBlobIds().get(0);
+      if (blobs.containsKey(key.getID())) {
+        ByteBuffer buf = blobs.get(key.getID()).duplicate();
+        buf.getLong(); // read off the size
+        buf.getShort(); // read off the type.
+        PutRequest originalBlobPutReq =
+            PutRequest.readFrom(new DataInputStream(new ByteBufferInputStream(buf)), clusterMap);
+        switch (getRequest.getMessageFormatFlag()) {
+          case BlobInfo:
```

Comment:
even though uncalled for, just leave a comment here, for any changes in MessageFormatRecord, this will break(i mean, tests depending on this will fail)